File: project/src/main/java/br/com/project/security/JwtAuthenticationFilter.java
package br.com.project.security;

import br.com.project.models.UserLogin;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Date;

import static br.com.project.security.SecurityConstants.*;

public class JwtAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

    private final AuthenticationManager authenticationManager;

    public JwtAuthenticationFilter(AuthenticationManager authenticationManager) {
        this.authenticationManager = authenticationManager;
    }

    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {
        try{
            // busca as credenciais digitadas pelo usuário
            UserLogin u = new ObjectMapper().readValue(request.getInputStream(), UserLogin.class);

            UsernamePasswordAuthenticationToken authResult = new UsernamePasswordAuthenticationToken(
                    u.getUsername(), u.getPassword()
            );

            // realiza a autenticação
            return authenticationManager.authenticate(authResult);

        } catch(IOException | AuthenticationException e) {
            throw new RuntimeException("Fail to authenticate!");
        }
    }

    @Override
    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) throws IOException, ServletException {
        String username = ((UserDetails) authResult.getPrincipal()).getUsername();

        // criação do token
        String jwtToken = JWT.create()
                .withSubject(username)
                .withExpiresAt(new Date(System.currentTimeMillis() + EXPIRATION))
                .sign(Algorithm.HMAC512(SECRET.getBytes()));

        String headerBody = username + " " + jwtToken;

        response.setHeader(HttpHeaders.AUTHORIZATION , headerBody);
    }
}

File: src/main/kotlin/br/com/project/projetoIntegrador/security/JwtAuthenticationFilter.kt
package br.com.project.projetoIntegrador.security

import br.com.project.projetoIntegrador.models.UserLogin
import br.com.project.projetoIntegrador.security.SecurityConstants.EXPIRATION
import br.com.project.projetoIntegrador.security.SecurityConstants.SECRET
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import java.util.*

class JwtAuthenticationFilter(
    private val authManager: AuthenticationManager
    ): UsernamePasswordAuthenticationFilter() {

    override fun attemptAuthentication(
        request: HttpServletRequest,
        response: HttpServletResponse
    ): Authentication {
        val credentials = ObjectMapper().readValue(request.inputStream, UserLogin::class.java)
        val authResult = UsernamePasswordAuthenticationToken(
            credentials.username,
            credentials.password,
            Collections.emptyList()
        )

        return authManager.authenticate(authResult)
    }

    override fun successfulAuthentication(
        request: HttpServletRequest?,
        response: HttpServletResponse,
        chain: FilterChain?,
        authResult: Authentication
    ) {
        val username = (authResult.principal as UserSecurity).username

        val jwtToken: String = JWT.create()
            .withSubject(username)
            .withExpiresAt(Date(System.currentTimeMillis() + EXPIRATION))
            .sign(Algorithm.HMAC512(SECRET.encodeToByteArray()))

        val headerBody = "Bearer $jwtToken"

        response.addHeader("Authorization", headerBody)
    }

    override fun unsuccessfulAuthentication(
        request: HttpServletRequest,
        response: HttpServletResponse,
        failed: AuthenticationException
    ) {
        throw BadCredentialsException("Email e/ou senha inválidos!")
    }
}